pkg/util/helper/unstructured.go

```diff
@@ -62,5 +63,12 @@ func ToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &unstructured.Unstructured{Object: uncastObj}, nil
+	unstructuredObj := &unstructured.Unstructured{Object: uncastObj}
+	if unstructuredObj.GetKind() == "" {
+		return nil, runtime.NewMissingKindErr(klog.KObj(unstructuredObj).String())
+	}
+	if unstructuredObj.GetAPIVersion() == "" {
+		return nil, runtime.NewMissingVersionErr(klog.KObj(unstructuredObj).String())
+	}
```

Comment:
I still can't understand why this `ToUnstructured` needs to be so picky, requiring apiVersion/kind to be set. In other words, what problem would arise without this? 

Basically, I'm asking `why we need it`, note that missing apiVersion/kind is not `ToUnstructured`'s fault, it just do the conversion job.

Reply:
> note that missing apiVersion/kind is not ToUnstructured's fault, it just do the conversion job.

I thought so too. But the problem is that the root cause of this kind of error is too obscure, and even if I fix that test, there is no way to prevent the next person from writing the next invalid test. 
From the perspective of preventing it from happening, if there is a way to completely prevent such errors from happening, I would welcome it no matter what it is, so I chose to modify this function, which seems to be the lowest cost solution.

Reply:
> I thought so too. But the problem is that the root cause of this kind of error is too obscure, and even if I fix that test, there is no way to prevent the next person from writing the next invalid test.

I agree. 
Another question, is doing all this just to prevent writing ineffective unit tests? Could business code potentially run into this issue due to oversight?

Reply:
> Another question, is doing all this just to prevent writing ineffective unit tests? Could business code potentially run into this issue due to oversight?

According to my statistics, currently only the test code is actually affected. 

Generally, I think new features are fully tested before being merged, so I don't think it is easy for production code to mix in problematic implementations. However, when writing https://github.com/karmada-io/karmada/pull/6076, this problem has puzzled me for a long time. In this PR, many `TypeMeta` additions are to ensure that https://github.com/karmada-io/karmada/pull/6076 can pass the test.

Reply:
> This issue blocks me to continue https://github.com/karmada-io/karmada/issues/6017.

So, this issue actually not a blocker of #6017, right?

Anyway, I will look at #6076 and get back to this, in order to understand the affection.

Reply:
Specifically, it‘s the tests involved in this PR that blocked #6076.